doc: Add LwM2M engine state machine (#45615)
zephyrproject-rtos/zephyr: doc/connectivity/networking/api/lwm2m.rst

For a more detailed LwM2M client sample see: :ref:`lwm2m-client-sample`.

LwM2M engine and application events
***********************************

> The Zephyr LwM2M engine defines events that can be sent back to the application via callback
> functions. The engine state machine shows when the events are spawned. Events depicted in
> the diagram are listed in the table. The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.

Reviewer: ```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
The engine state machine shows when the events are spawned.
Events depicted in the diagram are listed in the table.
The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```

.. figure:: images/lwm2m_engine_state_machine.png
    :alt: LwM2M engine state machine

    State machine for the LwM2M engine

.. list-table:: LwM2M RD Client events
   :widths: auto
   :header-rows: 1

   * - Event ID
     - Event Name
     - Description
     - Actions
   * - 0
     - NONE
     - No event
     - Do nothing
   * - 1
     - BOOTSTRAP_REG_FAILURE
     - Bootstrap registration failed.

>        Happens if timeout or failure in bootstrap registration.

Reviewer: ```suggestion
       Occurs if there is a timeout or failure in bootstrap registration.
```

     - Retry bootstrap
   * - 2
     - BOOTSTRAP_REG_COMPLETE
     - Bootstrap registration complete.

>        Happens after successful bootstrap registration.

Reviewer: ```suggestion
       Occurs after successful bootstrap registration.
```

     - No actions needed
   * - 3
     - BOOTSTRAP_TRANSFER_COMPLETE

>      - Bootstrap finish command received from the server.

Reviewer: ```suggestion
     - Bootstrap complete command received from the server.
```

Reviewer 2: The operation is really named `Bootstrap-Finish`
https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-1-7-2-0-6172-Bootstrap-Finish-Operation

     - No actions needed, client proceeds to registration.
   * - 4
     - REGISTRATION_FAILURE
     - Registration to LwM2M server failed.

>        Happens if timeout or failure in the registration.

Reviewer: ```suggestion
       Occurs if there is a timeout or failure in the registration.
```

     - Retry registration
   * - 5
     - REGISTRATION_COMPLETE
     - Registration to LwM2M server successful.

>        Happens after a successful registration reply from LwM2M server

Reviewer: ```suggestion
       Occurs after a successful registration reply from the LwM2M server
```

       or when session resumption is used.
     - No actions needed
   * - 6
     - REG_UPDATE_FAILURE
     - Registration update failed.

>        Happens if a timeout happens during registration update.

Reviewer: ```suggestion
       Occurs if a there is a timeout during registration update.
```

       NOTE: If registration update fails without a timeout,
       a full registration is triggered automatically and
       no registration update failure event is generated.
     - No actions needed, client proceeds to re-registration automatically.
   * - 7
     - REG_UPDATE_COMPLETE
     - Registration update completed.

>        Happens after successful registration update reply from LwM2M server.

Reviewer: ```suggestion
       Occurs after successful registration update reply from the LwM2M server.
```

     - No actions needed
   * - 8
     - DEREGISTER_FAILURE
     - Deregistration to LwM2M server failed.
       Happens if timeout or failure in the deregistration.
     - No actions needed, client proceeds to idle state automatically.
   * - 9
     - DISCONNECT
     - Disconnected from LwM2M server.

>        Happens if timeout during discussing with server.

Reviewer: ```suggestion
       Occurs if there is a timeout during the dialogue with the server.
```

Reviewer 2: Maybe this should be `communication with the server`

       Also triggered after deregistration has been done.

>      - No actions needed

Reviewer 2: Should there be a mention that if connection is required, application should restart the client.

For a more detailed LwM2M client sample see: :ref:`lwm2m-client-sample`.

LwM2M engine and application events
***********************************

> The Zephyr LwM2M engine defines events that can be sent back to the application via callback
> functions.

Reviewer: ```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
```

The engine state machine shows when the events are spawned.
Events depicted in the diagram are listed in the table.

> The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.

Reviewer: ```suggestion
The events are prefixed with ``LWM2M_RD_CLIENT_EVENT_``.
```
